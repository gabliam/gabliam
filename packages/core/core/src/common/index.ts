export * from './utils';
export * from './interfaces';
