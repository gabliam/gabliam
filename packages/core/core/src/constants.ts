/** Type in registry */
export const TYPE = {
  Service: 'ServiceType',
  Config: 'ConfigType',
  PreDestroy: 'PreDestroy',
};

export const METADATA_KEY = {
  bean: '_bean',
  onMissingBean: '_onMissingBean',
  init: '_init',
  beforeCreate: '_beforeCreate',
  value: '_value',
  register: '_register',
  scan: '_scan',
  config: '_config',
  service: '_service',
  plugin: '_plugin',
  injectContainer: '_inject_container',
  preDestroy: '_preDestroy',
};

/**
 * Inject container key in descriptor
 */
export const INJECT_CONTAINER_KEY = '$$container';

export const ORDER_CONFIG = {
  Core: 50,
  Plugin: 100,
  Config: 150,
};

export const ERRORS_MSGS = {
  DUPLICATED_DECORATOR: `Cannot apply decorator multiple times.`,
  DUPLICATED_CONFIG_DECORATOR: `Cannot apply @Config, @PluginConfig or @CorePlugin decorator multiple times.`,
  DUPLICATED_PLUGIN_DECORATOR: `Cannot apply @Plugin decorator multiple times.`,
  DUPLICATED_APPLICATION_DECORATOR: `Cannot apply @Application decorator multiple times.`,
};

export const APP_CONFIG = Symbol('APP_CONFIG');

export const CORE_CONFIG = Symbol('CORE_CONFIG');

export const VALUE_EXTRACTOR = Symbol('ValueExtractor');
