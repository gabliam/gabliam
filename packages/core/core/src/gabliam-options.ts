import { GabliamConfig } from './interfaces';
import * as _ from 'lodash';

const DEFAULT_CONFIG: GabliamConfig = {
  config: process.env.GABLIAM_CONFIG_PATH,
};

export const getGabliamConfig = (
  options?: Partial<GabliamConfig> | string
): GabliamConfig => {
  if (options === undefined) {
    return DEFAULT_CONFIG;
  } else {
    if (_.isString(options)) {
      return {
        ...DEFAULT_CONFIG,
        config: options,
      };
    } else {
      return {
        ...DEFAULT_CONFIG,
        ...options,
      };
    }
  }
};
