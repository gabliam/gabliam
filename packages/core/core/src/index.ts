import 'reflect-metadata';

export {
  decorate,
  inject,
  injectable,
  interfaces as inversifyInterfaces,
  multiInject,
  named,
  optional,
  postConstruct,
  tagged,
  targetName,
  unmanaged,
} from 'inversify';
export * from './common';
export {
  APP_CONFIG,
  CORE_CONFIG,
  INJECT_CONTAINER_KEY,
  VALUE_EXTRACTOR,
} from './constants';
export * from './container';
export * from './decorator';
export * from './errors';
export * from './gabliam';
export * from './gabliam-options';
export * from './gabliam-utils';
export * from './interfaces';
export * from './joi';
export { configResolver, FileLoader, Loader, Resolver } from './loaders';
export * from './metadatas';
export * from './reflection';
export * from './registry';
export * from './value-extractor';
