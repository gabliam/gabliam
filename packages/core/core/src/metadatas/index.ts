export * from './bean';
export { Config, PluginConfig } from './config';
export * from './init';
export * from './inject-container';
export * from './lifecycle';
export * from './plugin';
export * from './register';
export * from './scan';
export * from './service';
export * from './value';
