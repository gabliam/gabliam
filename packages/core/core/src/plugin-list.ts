import Graph from 'graph-data-structure';
import _ from 'lodash';
import { InvalidPluginError, PluginDependencyIsMissingError } from './errors';
import {
  GabliamAddPlugin,
  GabliamPlugin,
  GabliamPluginConstructor,
  GabliamPluginDefinition,
  GabliamPluginWithBind,
  GabliamPluginWithBuild,
  GabliamPluginWithConfig,
  GabliamPluginWithDestroy,
  GabliamPluginWithStart,
  GabliamPluginWithStop,
  PluginDependency,
} from './interfaces';
import { Plugin } from './metadatas';
import { reflection } from './reflection';

const getPluginDefinition = (
  definition: GabliamAddPlugin,
): GabliamPluginDefinition => {
  let pluginMetadata = reflection.annotationsOfDecorator<Plugin>(
    definition,
    Plugin,
  );

  let ctor: GabliamPluginConstructor | undefined;
  let plugin: GabliamPlugin | undefined;
  // if class doesn't have plugin metadata, so throw error
  if (pluginMetadata.length === 0 && definition.constructor) {
    pluginMetadata = reflection.annotationsOfDecorator<Plugin>(
      definition.constructor,
      Plugin,
    );
    if (pluginMetadata.length !== 0) {
      plugin = <GabliamPlugin>definition;
      ctor = <GabliamPluginConstructor>definition.constructor;
    }
  } else {
    ctor = <GabliamPluginConstructor>definition;
    // eslint-disable-next-line new-cap
    plugin = new ctor();
  }

  if (ctor === undefined || plugin === undefined) {
    throw new InvalidPluginError();
  }

  const [{ name: pluginName, beforeAll }] = (
    reflection.annotationsOfDecorator<Plugin>(ctor, Plugin, false) || []
  ).slice(-1);

  const name = pluginName || ctor.name;

  const dependencies = pluginMetadata.reduce<PluginDependency[]>(
    (prev, current) => {
      prev.push(...current.dependencies);
      return prev;
    },
    [],
  );

  return { plugin, name, dependencies, beforeAll };
};

/**
 * Plugin registry
 */
export class PluginList {
  private _plugins: GabliamPluginDefinition[] = [];

  /**
   * return list of plugin
   */
  get plugins(): GabliamPlugin[] {
    // use this._plugins.map for immutability
    return this._plugins.map((p) => p.plugin);
  }

  /**
   * return plugin with build phase
   */
  get pluginsWithBuild(): GabliamPluginWithBuild[] {
    return <GabliamPluginWithBuild[]>(
      this.plugins.filter((plugin) => _.isFunction(plugin.build))
    );
  }

  /**
   * return plugin with start phase
   */
  get pluginsWithStart(): GabliamPluginWithStart[] {
    return <GabliamPluginWithStart[]>(
      this.plugins.filter((plugin) => _.isFunction(plugin.start))
    );
  }

  /**
   * return plugin with stop phase
   */
  get pluginsWithStop(): GabliamPluginWithStop[] {
    return <GabliamPluginWithStop[]>(
      this.plugins.filter((plugin) => _.isFunction(plugin.stop))
    );
  }

  /**
   * return plugin with destroy phase
   */
  get pluginsWithDestroy(): GabliamPluginWithDestroy[] {
    return <GabliamPluginWithDestroy[]>(
      this.plugins.filter((plugin) => _.isFunction(plugin.destroy))
    );
  }

  /**
   * return plugin with bind phase
   */
  get pluginsWithBind(): GabliamPluginWithBind[] {
    return <GabliamPluginWithBind[]>(
      this.plugins.filter((plugin) => _.isFunction(plugin.bind))
    );
  }

  /**
   * return plugin with config phase
   */
  get pluginWithConfig(): GabliamPluginWithConfig[] {
    return <GabliamPluginWithConfig[]>(
      this.plugins.filter((plugin) => _.isFunction(plugin.config))
    );
  }

  /**
   * Add a plugin
   */
  add(definition: GabliamAddPlugin): GabliamPluginDefinition {
    const def = getPluginDefinition(definition);
    this._plugins.push(def);
    return def;
  }

  /**
   * Sort plugin with the good order
   */
  sort() {
    const graph = Graph();
    const beforeAll = [];
    for (const plugin of this._plugins) {
      if (plugin.beforeAll) {
        beforeAll.push(plugin.name);
      }
    }

    for (const plugin of this._plugins) {
      graph.addNode(plugin.name);

      beforeAll.forEach((def) => graph.addEdge(def, plugin.name));

      // const orderedPlugin = [plugin.name];
      if (plugin.dependencies) {
        for (const deps of plugin.dependencies) {
          if (typeof deps.name !== 'string') {
            const def = this.add(deps.name);
            // eslint-disable-next-line default-case
            switch (deps.order) {
              case 'after':
                graph.addEdge(def.name, plugin.name);
                break;
              case 'before':
                graph.addEdge(plugin.name, def.name);
                break;
            }
          } else {
            if (!this.has(deps.name)) {
              throw new PluginDependencyIsMissingError(plugin.name, deps.name);
            }
            // eslint-disable-next-line default-case
            switch (deps.order) {
              case 'after':
                graph.addEdge(deps.name, plugin.name);
                break;
              case 'before':
                graph.addEdge(plugin.name, deps.name);
                break;
            }
          }
        }
      }
    }

    const listPlugin = graph
      .topologicalSort()
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      .map<GabliamPluginDefinition>((p) => this.findByName(p)!);

    this._plugins = listPlugin;
  }

  /**
   * Checks if `name` is in plugin list.
   */
  has(name: string) {
    return !!this.findByName(name);
  }

  /**
   * Find if `name` is in plugin list
   */
  findByName(name: string) {
    return this._plugins.find((p) => p.name === name);
  }
}
