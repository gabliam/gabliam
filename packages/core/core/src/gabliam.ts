/* eslint-disable no-await-in-loop */
import { ExpressionParser } from '@gabliam/expression';
import d from 'debug';
import _ from 'lodash';
import { callInstance, toPromise } from './common';
import { APP_CONFIG, CORE_CONFIG, TYPE, VALUE_EXTRACTOR } from './constants';
import { Container, createContainer } from './container';
import {
  ConfigRegistry,
  GabliamAddPlugin,
  GabliamConfig,
  PreDestroyRegistry,
  ValueRegistry,
} from './interfaces';
import { LoaderConfig, LoaderModule } from './loaders';
import {
  Bean,
  BeforeCreate,
  Init,
  OnMissingBean,
  PreDestroy,
} from './metadatas';
import { PluginList } from './plugin-list';
import { reflection } from './reflection';
import { Registry } from './registry';
import { configureValueExtractor } from './value-extractor';
import { getGabliamConfig } from './gabliam-options';

const debug = d('Gabliam:core');

/**
 * Gabliam
 */
export class Gabliam {
  public container: Container = createContainer();

  public config: object;

  /**
   * Registry
   */
  public registry: Registry = new Registry();

  /**
   * Module loader
   */
  public loaderModule: LoaderModule = new LoaderModule();

  /**
   * Config loader
   */
  public loaderConfig: LoaderConfig = new LoaderConfig();

  /**
   * Plugin list
   */
  public pluginList: PluginList = new PluginList();

  public options: GabliamConfig;

  /**
   * Constructor
   * @param  {GabliamConfig|string} options?
   */
  constructor(options?: Partial<GabliamConfig> | string) {
    this.options = getGabliamConfig(options);
    /**
     * @TODO move in building phase
     */
    this.container.bind(CORE_CONFIG).toConstantValue(this.options);
  }

  /**
   * Add a plugin
   * @param  {GabliamPluginConstructor} plugin
   * @returns Gabliam
   */
  public addPlugin(plugin: GabliamAddPlugin): Gabliam {
    this.pluginList.add(plugin);
    return this;
  }

  /**
   * Add any plugins
   * @param  {GabliamPluginConstructor[]} ...plugins
   * @returns Gabliam
   */
  public addPlugins(...plugins: Array<GabliamAddPlugin>): Gabliam {
    for (const plugin of plugins) {
      this.addPlugin(plugin);
    }
    return this;
  }

  /**
   * Build gabliam
   * @returns Promise
   */
  public async build(): Promise<Gabliam> {
    this.pluginList.sort();

    /**
     * Load config file
     */
    await this._initializeConfig();

    /**
     * Loading phase
     */
    this.registry.addRegistry(
      this.loaderModule.load(this.options.scanPath, this.pluginList.plugins),
    );

    /**
     * Binding phase
     */
    await this._bind();

    /**
     * Config phase
     */
    await this._loadConfig();

    for (const plugin of this.pluginList.pluginsWithBuild) {
      await toPromise(plugin.build(this.container, this.registry));
    }

    return this;
  }

  /**
   * Build and start gabliam application
   * @returns Promise
   */
  async buildAndStart(): Promise<Gabliam> {
    await this.build();
    await this.start();
    return this;
  }

  /**
   * Stop and destroy gabliam application
   * @returns Promise
   */
  async stopAndDestroy(): Promise<Gabliam> {
    await this.stop();
    await this.destroy();
    return this;
  }

  /**
   * Start gabliam application
   *
   * call all plugin.start
   * @returns Promise
   */
  async start(): Promise<Gabliam> {
    for (const plugin of this.pluginList.pluginsWithStart) {
      await toPromise(plugin.start(this.container, this.registry));
    }

    return this;
  }

  /**
   * Stop gabliam application
   * call all plugin.stop
   *
   * @returns Promise
   */
  async stop(): Promise<Gabliam> {
    for (const plugin of this.pluginList.pluginsWithStop) {
      await toPromise(plugin.stop(this.container, this.registry));
    }

    return this;
  }

  /**
   * Destroy gabliam application
   * call all plugin.destroy
   *
   * @returns Promise
   */
  async destroy(): Promise<Gabliam> {
    for (const plugin of this.pluginList.pluginsWithDestroy) {
      await toPromise(plugin.destroy(this.container, this.registry));
    }

    const values = this.registry.get<PreDestroyRegistry>(TYPE.PreDestroy);

    const instanceToDestroy = async (
      value: ValueRegistry<PreDestroyRegistry>,
    ) => {
      if (value.options) {
        const instance = this.container.get(value.id);
        for (const preDestroy of value.options.preDestroys) {
          await callInstance(instance, preDestroy);
        }
      }
    };

    const p = [];
    for (const value of values) {
      p.push(instanceToDestroy(value));
    }

    await Promise.all(p);

    return this;
  }

  /**
   * Load config file and bind result in APP_CONFIG
   */
  private async _initializeConfig() {
    this.config = await this.loaderConfig.load(this.options.config);
    const config = this.config;
    this.container.bind(APP_CONFIG).toConstantValue(config);
    this.container
      .bind(VALUE_EXTRACTOR)
      .toConstantValue(configureValueExtractor(this.container));
    this.container
      .bind(ExpressionParser)
      .toConstantValue(new ExpressionParser(config));
  }

  /**
   * Binding phase
   * Binding of config and service classes.
   * call all plugin.bind
   */
  private async _bind() {
    debug('_bind');

    this.registry
      .getAllAutoBind()
      .forEach(({ id, target }) =>
        this.container.bind(id).to(target).inSingletonScope(),
      );

    for (const plugin of this.pluginList.pluginsWithBind) {
      await toPromise(plugin.bind(this.container, this.registry));
    }
  }

  /**
   * Config phase
   */
  private async _loadConfig() {
    debug('_loadConfig');

    // Get all config classes in registry
    let configsRegistry = this.registry.get<ConfigRegistry>(TYPE.Config);

    debug('configsRegistry', configsRegistry);
    if (configsRegistry) {
      configsRegistry = _.sortBy(configsRegistry, 'options.order');

      for (const { id: configId } of configsRegistry) {
        // Get config instance
        const confInstance = this.container.get<object>(configId);
        const ctor = confInstance.constructor;

        // get all bean metadata in config classes
        const beanMetadatas = reflection.propMetadataOfDecorator<Bean>(
          ctor,
          Bean,
        );

        // get on missing bean metadata
        const onMissingBeanMetadatas = reflection.propMetadataOfDecorator<OnMissingBean>(
          ctor,
          OnMissingBean,
        );

        const beforeCreateMetas = Object.keys(
          reflection.propMetadataOfDecorator(ctor, BeforeCreate),
        );

        const initMetadas = Object.keys(
          reflection.propMetadataOfDecorator(ctor, Init),
        );

        // call all beforeCreate method if exist
        if (Array.isArray(beforeCreateMetas)) {
          // No promise.all and await because order of beans are important
          for (const metada of beforeCreateMetas) {
            await callInstance(confInstance, metada);
          }
        }

        // If config has bean metadata
        if (Object.keys(beanMetadatas).length) {
          // No promise.all and await because order of beans are important
          for (const [key, beans] of Object.entries(beanMetadatas)) {
            const onMissingBeans = onMissingBeanMetadatas[key] || [];

            // by default all bean are missing
            let allMissing = true;

            // if there are onMissingBeans config, check if bean exist or no
            for (const onMissingBean of onMissingBeans) {
              try {
                this.container.getAll(onMissingBean.id);
                allMissing = false;
                // eslint-disable-next-line no-empty
              } catch {}
            }

            // if all beans are missing, so we create all beans
            if (allMissing) {
              for (const { id } of beans) {
                const bean = await callInstance(confInstance, key);
                this.container.bind(id).toConstantValue(bean);
                if (bean === undefined || bean.constructor === undefined) {
                  // eslint-disable-next-line no-continue
                  continue;
                }

                const preDestroys = Object.keys(
                  reflection.propMetadataOfDecorator(
                    bean.constructor,
                    PreDestroy,
                  ),
                );

                // bean can return undefined or can be a constant value
                if (preDestroys.length) {
                  this.registry.add(TYPE.PreDestroy, <ValueRegistry>{
                    id,
                    target: bean,
                    options: {
                      preDestroys,
                    },
                  });
                }
              }
            }
          }
        }

        if (Array.isArray(initMetadas)) {
          // No promise.all and await because order of beans are important
          for (const metada of initMetadas) {
            await callInstance(confInstance, metada);
          }
        }

        for (const plugin of this.pluginList.pluginWithConfig) {
          await toPromise(
            plugin.config(this.container, this.registry, confInstance),
          );
        }
      }
    }
    debug('_loadConfig end');
  }
}
