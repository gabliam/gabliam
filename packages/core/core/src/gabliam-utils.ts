import * as bluebird from 'bluebird';
import * as g from 'glob';
import { toPromise, Type } from './common';
import { TYPE } from './constants';
import { Gabliam } from './gabliam';
import { ValueRegistry } from './interfaces';
import { LoaderModule } from './loaders';
import { Application } from './metadatas';
import { reflection } from './reflection';
import { Registry } from './registry';
import { promisify } from 'util';
import * as _ from 'lodash';
const glob = promisify(g);

const reg = /^.*(git|svn|node_modules|dist|build).*/;

/**
 * Search gabliam app and return array of valueregistry
 */
export const gabliamFinder = async (scanPath: string) => {
  const loaderModule = new LoaderModule();
  const registry = new Registry();
  registry.addRegistry(loaderModule.load(scanPath, []));
  return registry.get(TYPE.Application);
};

/**
 * Find gabliam application and return the class
 * @param scanPath path to scan
 * @param appName name of the application (if many)
 */
export const gabliamFindApp = async (scanPath: string, appName?: string) => {
  const list = await gabliamFinder(scanPath);
  let app: Type<any> | undefined;
  if (appName) {
    for (const value of list) {
      const metadata = reflection.annotationsOfDecorator<Application>(
        value.target,
        Application
      )[0];

      if (metadata) {
        const name = metadata.name ?? value.target.name;
        if (appName === name) {
          app = value.target;
          break;
        }
      }
    }
  } else {
    if (list.length > 1) {
      throw new Error(
        `Too many app (${list.length}) found. You must select an application.`
      );
    }

    if (list[0]) {
      app = list[0].target;
    }
  }

  if (app === undefined) {
    const errorMsg = appName
      ? `${appName} is not a gabliam application`
      : 'no application found';
    throw new Error(errorMsg);
  }

  return app;
};

const isValueRegistry = (val: any): val is ValueRegistry<any> => {
  return (
    val &&
    typeof val === 'object' &&
    val.hasOwnProperty('id') &&
    val.hasOwnProperty('target') &&
    val.hasOwnProperty('autoBind')
  );
};

/**
 * Build a gabliam app
 * @param clazzOrValue  ValueRegistry or class of application
 */
export const gabliamBuilder = async <T = any>(
  clazzOrValue: ValueRegistry<T> | Type<any>
) => {
  let clazz: any = clazzOrValue;
  if (isValueRegistry(clazzOrValue)) {
    clazz = clazzOrValue.target;
  }

  const application = reflection.annotationsOfDecorator<Application>(
    clazz,
    Application
  )[0];
  if (application) {
    const plugins = await bluebird
      .filter(application.plugins, async item => toPromise(item.condition()))
      .map(item => {
        if (typeof item.plugin === 'string') {
          return require(item.plugin).default;
        }
        return item.plugin;
      });
    return new Gabliam(application.gabliamConfig).addPlugins(...plugins);
  }

  throw new Error(`${clazzOrValue} is not a gabliam application`);
};

export const setupTsProject = async (folder: string) => {
  const files = await bluebird.filter(
    glob('{*.@(|ts),!(git|svn|node_modules|dist|build)/**/*.@(ts)}', {
      cwd: folder,
    }),
    file => !_.endsWith(file, '.d.ts') && !reg.test(file)
  );
  if (files.length > 0) {
    console.log('Typescript project detected. Add ts-node');
    const tsnode = require('ts-node');
    tsnode.register({
      dir: process.cwd(),
    });
    module.require('tsconfig-paths/register');
  }
};
