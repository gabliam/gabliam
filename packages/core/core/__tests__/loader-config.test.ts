// tslint:disable:one-line
// tslint:disable:no-unused-expression
import { LoaderConfig, FileLoader } from '../src/loaders';
// import * as mock from 'mock-fs';
import { MockFs } from './mock-fs';
import * as path from 'path';

let loader: LoaderConfig;
beforeEach(async () => {
  loader = new LoaderConfig();
});

test(`with config folder doesn't exist`, async () => {
  const config = await loader.load(__dirname, 'doesntexist/config');
  expect(config).toMatchSnapshot();
});

test(`with bad application.yml`, async () => {
  const config = loader.load(path.resolve(__dirname, './fixtures/badyml'));
  await expect(config).rejects.toMatchSnapshot();
});

test(`with application.yml empty`, async () => {
  const mock = new MockFs({
    'test/config': {
      'application.yml': ``,
    },
  });
  const config = await loader.load(mock.resolve('test/config'));
  expect(config).toMatchSnapshot();
  mock.restore();
});

test(`with application.yml with nothing`, async () => {
  const mock = new MockFs({
    'test/config': {
      // tslint:disable-next-line:no-trailing-whitespace
      'application.yml': `

              `,
    },
  });
  const config = await loader.load(mock.resolve('test/config'));
  expect(config).toMatchSnapshot();
  mock.restore();
});

test(`with application.yml`, async () => {
  const config = await loader.load(
    path.resolve(__dirname, './fixtures/config')
  );
  expect(config).toMatchSnapshot();
});

describe('with application.yml', () => {
  let mock: MockFs;
  beforeAll(async () => {
    mock = new MockFs({
      'test/config': {
        'application.yml': `
              application:
                host: 127.0.0.1
                db: test
              testPath: path:./test.json
            `,
        'application-prod.yml': `
              application:
                host: prod.app.com
            `,
      },
    });
  });

  test('load application.yml', async () => {
    const config = await loader.load(mock.resolve('test/config'));
    expect(config).toMatchSnapshot();
  });

  test('load with bad profile', async () => {
    const config = await loader.load(mock.resolve('test/config'), 'int');
    expect(config).toMatchSnapshot();
  });

  test('load with prod profile', async () => {
    const config = await loader.load(mock.resolve('test/config'), 'prod');
    expect(config).toMatchSnapshot();
  });

  afterAll(() => mock.restore());
});

describe('with application.yml with bad handler', () => {
  let mock: MockFs;
  beforeAll(async () => {
    mock = new MockFs({
      'test/config': {
        'application.yml': `
              application:
                host: 127.0.0.1
                db: test
              testPath: file:./test.json
            `,
        'application-prod.yml': `
              application:
                host: prod.app.com
            `,
      },
    });
  });

  test('must throw error', async () => {
    const config = loader.load(mock.resolve('test/config'), 'int');
    await expect(config).rejects.toMatchSnapshot();
  });

  afterAll(() => mock.restore());
});

test(`with application.json`, async () => {
  const config = await loader.load([
    {
      loader: FileLoader,
      options: {
        folder: path.resolve(__dirname, './fixtures/config'),
        types: ['json'],
      },
    },
  ]);
  expect(config).toMatchSnapshot();
});

test(`with json and yml`, async () => {
  const config = await loader.load([
    {
      loader: FileLoader,
      options: {
        folder: path.resolve(__dirname, './fixtures/config'),
        types: ['json', 'yml'],
      },
    },
  ]);
  expect(config).toMatchSnapshot();
});

test(`with yml and json`, async () => {
  const config = await loader.load([
    {
      loader: FileLoader,
      options: {
        folder: path.resolve(__dirname, './fixtures/config'),
        types: ['yml', 'json'],
      },
    },
  ]);
  expect(config).toMatchSnapshot();
});

test(`with bad application.json`, async () => {
  const config = loader.load([
    {
      loader: FileLoader,
      options: {
        folder: path.resolve(__dirname, './fixtures/badjson'),
        types: ['json'],
      },
    },
  ]);
  await expect(config).rejects.toMatchSnapshot();
});

test(`with application.json empty`, async () => {
  const mock = new MockFs({
    'test/config': {
      'application.json': ``,
    },
  });
  const config = loader.load([
    {
      loader: FileLoader,
      options: {
        folder: mock.resolve('test/config'),
        types: ['json'],
      },
    },
  ]);
  await expect(config).rejects.toMatchSnapshot();
  mock.restore();
});

test(`with application.json with nothing`, async () => {
  const mock = new MockFs({
    'test/config': {
      // tslint:disable-next-line:no-trailing-whitespace
      'application.json': `

              `,
    },
  });
  const config = loader.load([
    {
      loader: FileLoader,
      options: {
        folder: mock.resolve('test/config'),
        types: ['json'],
      },
    },
  ]);
  await expect(config).rejects.toMatchSnapshot();
  mock.restore();
});

test(`with bad parser`, async () => {
  const config = loader.load([
    {
      loader: FileLoader,
      options: {
        folder: path.resolve(__dirname, './fixtures/badparser'),
        types: ['badparser'],
      },
    },
  ]);
  await expect(config).rejects.toMatchSnapshot();
});

test(`with bad loader`, async () => {
  const config = loader.load([
    {
      loader: 'BadLoader',
    },
  ]);
  await expect(config).rejects.toMatchSnapshot();
});

test(`with toml`, async () => {
  const mock = new MockFs({
    'test/config': {
      // tslint:disable-next-line:no-trailing-whitespace
      'application.toml': `
[application]
host = "127.0.0.1"
db = "test"

testPath = "path:./test.json"`,
    },
  });
  const config = await loader.load([
    {
      loader: FileLoader,
      options: {
        folder: mock.resolve('test/config'),
        types: ['toml'],
      },
    },
  ]);
  expect(config).toMatchSnapshot();
  mock.restore();
});

test(`with bad toml`, async () => {
  const mock = new MockFs({
    'test/config': {
      // tslint:disable-next-line:no-trailing-whitespace
      'application.toml': `
;;;;[application]
host = "127.0.0.1"
db = "test"

testPath = "path:./test.json"`,
    },
  });
  const config = loader.load([
    {
      loader: FileLoader,
      options: {
        folder: mock.resolve('test/config'),
        types: ['toml'],
      },
    },
  ]);
  await expect(config).rejects.toMatchSnapshot();
  mock.restore();
});

test(`with properties`, async () => {
  const mock = new MockFs({
    'test/config': {
      // tslint:disable-next-line:no-trailing-whitespace
      'application.properties': `
application.host = 127.0.0.1
application.db = test
testPath = path:./test.json`,
    },
  });
  const config = await loader.load([
    {
      loader: FileLoader,
      options: {
        folder: mock.resolve('test/config'),
        types: ['properties'],
      },
    },
  ]);
  expect(config).toMatchSnapshot();
  mock.restore();
});

test.skip(`with bad properties`, async () => {
  const mock = new MockFs({
    'test/config': {
      // tslint:disable-next-line:no-trailing-whitespace
      'application.properties': `
    ####😁😁$$$[application]
      host = "127.0.0.1"
db = "test"

testPath = "path:./test.json"`,
    },
  });
  const config = loader.load([
    {
      loader: FileLoader,
      options: {
        folder: mock.resolve('test/config'),
        types: ['properties'],
      },
    },
  ]);
  await expect(config).rejects.toMatchSnapshot();
  mock.restore();
});
