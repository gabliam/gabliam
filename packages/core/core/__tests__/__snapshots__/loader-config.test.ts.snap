// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with application.json 1`] = `
Object {
  "application": Object {
    "host": "prod.app.json",
    "json": true,
    "mongoose": Object {
      "mongoose": Object {
        "url": "mongo://localhost",
      },
    },
  },
}
`;

exports[`with application.json empty 1`] = `
[LoaderConfigParseError: Error during parsing of file "<TEMP_DIR>/mock-fs-test/test/config/application.json".
    SyntaxError: Unexpected end of JSON input
    ]
`;

exports[`with application.json with nothing 1`] = `
[LoaderConfigParseError: Error during parsing of file "<TEMP_DIR>/mock-fs-test/test/config/application.json".
    SyntaxError: Unexpected end of JSON input
    ]
`;

exports[`with application.yml 1`] = `
Object {
  "application": Object {
    "host": "prod.app.yml",
    "mongoose": Object {
      "mongoose": Object {
        "url": "mongo://localhost",
      },
    },
    "yml": true,
  },
}
`;

exports[`with application.yml empty 1`] = `Object {}`;

exports[`with application.yml load application.yml 1`] = `
Object {
  "application": Object {
    "db": "test",
    "host": "127.0.0.1",
  },
  "testPath": "<TEMP_DIR>/mock-fs-test/test/config/test.json",
}
`;

exports[`with application.yml load with bad profile 1`] = `
Object {
  "application": Object {
    "db": "test",
    "host": "127.0.0.1",
  },
  "testPath": "<TEMP_DIR>/mock-fs-test/test/config/test.json",
}
`;

exports[`with application.yml load with prod profile 1`] = `
Object {
  "application": Object {
    "db": "test",
    "host": "prod.app.com",
  },
  "testPath": "<TEMP_DIR>/mock-fs-test/test/config/test.json",
}
`;

exports[`with application.yml with bad handler must throw error 1`] = `[Error: ENOENT: no such file or directory, open '<TEMP_DIR>/mock-fs-test/test/config/test.json']`;

exports[`with application.yml with nothing 1`] = `Object {}`;

exports[`with bad application.json 1`] = `
[LoaderConfigParseError: Error during parsing of file "<PROJECT_ROOT>/packages/core/core/__tests__/fixtures/badjson/application.json".
    SyntaxError: Unexpected token } in JSON at position 10
    ]
`;

exports[`with bad application.yml 1`] = `
[LoaderConfigParseError: Error during parsing of file "<PROJECT_ROOT>/packages/core/core/__tests__/fixtures/badyml/application.yml".
    YAMLException: can not read a block mapping entry; a multiline key may not be an implicit key at line 7, column 1:
    
    ^
    ]
`;

exports[`with bad loader 1`] = `[LoaderConfigPgkNotInstalledError: BadLoader package has not been found installed. Try to install it: npm install BadLoader --save or yarn add BadLoader]`;

exports[`with bad parser 1`] = `[ParserNotSupportedError: The parser "badparser" is not supported by Gabliam]`;

exports[`with bad toml 1`] = `
[LoaderConfigParseError: Error during parsing of file "<TEMP_DIR>/mock-fs-test/test/config/application.toml".
    TomlError: Unknown character "59" at row 2, col 1, pos 2:
1: 
2> ;;;;[application]
   ^
3: host = "127.0.0.1"


    ]
`;

exports[`with config folder doesn't exist 1`] = `Object {}`;

exports[`with json and yml 1`] = `
Object {
  "application": Object {
    "host": "prod.app.yml",
    "json": true,
    "mongoose": Object {
      "mongoose": Object {
        "url": "mongo://localhost",
      },
    },
    "yml": true,
  },
}
`;

exports[`with properties 1`] = `
Object {
  "application.db": "test",
  "application.host": "127.0.0.1",
  "testPath": "<TEMP_DIR>/mock-fs-test/test/config/test.json",
}
`;

exports[`with toml 1`] = `
Object {
  "application": Object {
    "db": "test",
    "host": "127.0.0.1",
    "testPath": "<TEMP_DIR>/mock-fs-test/test/config/test.json",
  },
}
`;

exports[`with yml and json 1`] = `
Object {
  "application": Object {
    "host": "prod.app.json",
    "json": true,
    "mongoose": Object {
      "mongoose": Object {
        "url": "mongo://localhost",
      },
    },
    "yml": true,
  },
}
`;
